src/state.rs

```diff
@@ -333,20 +570,17 @@ impl DemandProxyState {
         let workload_uid = workload.uid.to_owned();
         let hostname = workload.hostname.to_owned();
         metrics.as_ref().on_demand_dns.get_or_create(&labels).inc();
-        let rdns = match state.get_ips_for_hostname(&workload.hostname) {
+        let rdns = match state.dns_resolver.get_ips_for_hostname(&hostname) {
```

Comment:
Can this entire method (load_balance_for_hostname) just be moved to `DnsResolver`? All of this should be internal details of it, I think.

Reply:
Hi @nmittler, as you know `DnsResolver` is part of `DemandProxyState`, and it needs to modify itself on its 2 HashMaps, so it's mutable reference. If moving `load_balance_for_hostname` into `DnsResolver`, it would impact that the whole `DemandProxyState` need to be a mutable reference as well. So hope it can be as it is now. Does it make sense?

Reply:
I was actually concerned when I saw that [DemandProxyState.resolve_on_demand_dns](https://github.com/istio/ztunnel/blob/1b20547a865b70f90492ffb18cbe3fb86e3beea8/src/state.rs#L368) accepting a non-reference `self` just so that it can switch to a mutable reference. Not sure when that was introduced, but it seems a bit hacky and should probably be fixed if possible. Perhaps the issue is that we have some methods that needlessly require `&mut self`(e.g. if you lock an object that you write to, you don't need mut)?

@howardjohn I may be off here, since my rust is a little rusty at this point :) ... thoughts?